### PR TITLE
WIP: Check if minikube is already running

### PR DESCRIPTION
**Pull request by ravsa**
Fixes: #2646

### CODE REVIEW

**Comment by tstromberg** [comment]
@minikube-bot OK to test

**Comment by dlorenc** [comment]
Can you explain the purpose of this change? Start is supposed to be safe to call repeatedly.

**Comment by ravsa** [comment]
@dlorenc the purpose of this PR is to fix the issue  https://github.com/kubernetes/minikube/issues/2646 and possibly related issue https://github.com/kubernetes/minikube/issues/3284. IMHO it's good to skip the start process if it's already running.

**Comment by balopat** [comment]
@minikube-bot OK to test

**Comment by balopat** [comment]
@minikube-bot OK to test

**Comment by balopat** [comment]
@minikube-bot OK to test

**Comment by ravsa** [comment]
@balopat, could you help me with this build issue? it's taking too long.

**Comment by balopat** [comment]
@minikube-bot OK to test

File: cmd/minikube/cmd/start.go [hunk 1]
```diff
@@ -112,6 +115,7 @@ func runStart(cmd *cobra.Command, args []string) {
 	}
 
 	api, err := machine.NewAPIClient()
+	ensureNotRunning(api, constants.DefaultMachineName)
```
Comment (alexsomesan) [inline cmd/minikube/cmd/start.go]: What happens if the already running instance has a different name? (like via the `--profile` argument to `minikube start`)

Comment (ravsa) [reply cmd/minikube/cmd/start.go]: nice catch, I will make the changes. Thanks

**Comment by tstromberg** [comment]
@minikube-bot OK to test

**Comment by tstromberg** [comment]
@minikube-bot OK to test

**Comment by tstromberg** [comment]
I believe this PR breaks the workflow where folks make changes to their local configuration, such as updating the version of Kubernetes used, and use `minikube start` to update their running VM to reflect that configuration.

I'm tempted to close this PR - but perhaps I'm not interpreting this change correctly?

**Comment by tstromberg** [comment]
I've been thinking more about this, and have a few thoughts:

- Folks currently rely on 'minikube start'  to upgrade their VM's or apply small configuration changes.
- On the other hand, most folks just want fast access to their VM.

We need to bridge the gap between the two, and I think the only way to do so is to compare the old configuration against the new one. If they match, exit early, otherwise, go through the full process.

Any opinions here?

**Comment by stafot** [comment]
@tstromberg A good idea is to have 2 different "commands"/"flags" to serve these conflicting needs f.e a `minikube start-fast` vs a `minikube start-update`. Is it so crucial in your opinion to be unified under a same `start command` for keeping it simple? 
Although I think that you are raising a valid concern.

**Comment by tstromberg** [comment]
> @tstromberg A good idea is to have 2 different "commands"/"flags" to serve these conflicting needs f.e a `minikube start-fast` vs a `minikube start-update`. Is it so crucial in your opinion to be unified under a same `start command` for keeping it simple?
> Although I think that you are raising a valid concern.

How about adding a `--update` or `--reconfigure` flag to start?

In an ideal world, minikube should be able to detect whether or not reconfiguration is required, but it doesn't yet (PR's welcome!). This flag would allow users to override that detection, and specify whether or not to return quickly - as your PR does - or to forcibly reconfigure. 

For your solution, you would add `minikube start --reconfigure=false` or `--update=false`. Thoughts?

**Comment by tstromberg** [comment]
I'm closing as this PR seems to be inactive, and having unmergeable PR's around slows us down from making progress on the PR's that are active. Please re-open this PR once you've had a chance to address the feedback on it.

Rather than adding extra knobs, my preference is for this PR to check if the requested configuration is different than the previously setup configuration. If it's different, take the slow path - if it's the same, take the fast path. 

Thank you for proposing this PR. It has sparked a very important discussion.

**Pull request closed without merging**